plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/AddFilesProcedure.java

```diff
@@ -0,0 +1,404 @@
+
+        SchemaTableName targetName = new SchemaTableName(targetSchemaName, targetTableName);
+        io.trino.metastore.Table targetTable = metastore.getTable(targetSchemaName, targetTableName).orElseThrow(() -> new TableNotFoundException(targetName));
+        if (!isIcebergTable(targetTable)) {
+            throw new TrinoException(NOT_SUPPORTED, "The target table must be Iceberg table");
+        }
+        accessControl.checkCanInsertIntoTable(null, targetName);
+
+        Table table = catalog.loadTable(session, targetName);
+        PartitionSpec partitionSpec = table.spec();
+        Schema schema = table.schema();
+        NameMapping nameMapping = MappingUtil.create(schema);
+        Set<Integer> requiredFields = schema.columns().stream()
+                .filter(Types.NestedField::isRequired)
+                .map(Types.NestedField::fieldId)
+                .collect(toImmutableSet());
+
+        HiveStorageFormat storageFormat;
+        String sourceLocation;
+        io.trino.metastore.Table sourceTable = null;
+        SourceType sourceType;
+        if (sourceSchemaName != null && sourceTableName != null) {
+            checkProcedureArgument(location == null, "LOCATION argument must be null");
+            checkProcedureArgument(format == null, "FORMAT argument must be null");
+
+            SchemaTableName sourceName = new SchemaTableName(sourceSchemaName, sourceTableName);
+            sourceTable = metastore.getTable(sourceSchemaName, sourceTableName).orElseThrow(() -> new TableNotFoundException(sourceName));
+
+            accessControl.checkCanSelectFromColumns(null, sourceName, Stream.concat(sourceTable.getDataColumns().stream(), sourceTable.getPartitionColumns().stream())
+                    .map(Column::getName)
+                    .collect(toImmutableSet()));
+
+            checkProcedureArgument(
+                    table.schemas().size() == sourceTable.getDataColumns().size(),
+                    "Data column count mismatch: %d vs %d", table.schemas().size(), sourceTable.getDataColumns().size());
+            for (Column sourceColumn : sourceTable.getDataColumns()) {
+                Types.NestedField targetColumn = schema.caseInsensitiveFindField(sourceColumn.getName());
+                if (targetColumn == null) {
+                    throw new TrinoException(COLUMN_NOT_FOUND, "Column '%s' does not exist".formatted(sourceColumn.getName()));
+                }
+                ColumnIdentity columnIdentity = createColumnIdentity(targetColumn);
+                org.apache.iceberg.types.Type sourceColumnType = toIcebergType(typeManager.getType(getTypeSignature(sourceColumn.getType(), DEFAULT_PRECISION)), columnIdentity);
+                if (!targetColumn.type().equals(sourceColumnType)) {
+                    throw new TrinoException(TYPE_MISMATCH, "Expected target '%s' type, but got source '%s' type".formatted(targetColumn.type(), sourceColumnType));
+                }
+            }
```

Comment:
These checks may be a little strict, if Iceberg supports coercion from the source column to the target type it should be okay as is. 

Having additional columns in the source than the target also doesn't hurt, we'll just ignore them.

It is harder to mess up a call to the procedure this way, but I guess it's a question of how we expect people to use this. With perfectly matching schemas, or ones that have been tweaked a little.

Reply:
I think we should start with strict checks and tweak them based on user feedback. 

The common use case looks adding files from Hive partitions after the initial migration. For instance, migrating the entire table on July 24, adding files from July 25 partition next day, .... until the end of migration.

Reply:
Does Spark's version of add_files enforce strict checks, or is it permissive?

I'm also in favor of defaulting to strict mode, but giving the user the ability disable it with a `non-strict` boolean flag.

Reply:
https://github.com/trinodb/trino/issues/23677